// @framer/ruucm-h.lazy-load-image/code/LazyLoadImage.tsx
import * as React from "react";
import useLazyImg from "./use-lazy-img";
import { addPropertyControls, ControlType } from "framer";
import styled, { css } from "styled-components";

const BluredImage = styled.img`
  position: absolute;
  top: 0;
  /* animate */
  opacity: 1;
  ${({ loading, animateDuration }) =>
    !loading &&
    css`
      opacity: 0;
      transition: ${`${animateDuration}s opacity ease-in-out`};
    `}
`;

export function LazyLoadImage({
  imgUrl,
  placeholderUrl = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAAaADAAQAAAABAAAAAQAAAAD5Ip3+AAAADUlEQVQIHWPYsWPHfwAHfAMoI6JxIgAAAABJRU5ErkJggg==",
  fallbackUrl = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAAaADAAQAAAABAAAAAQAAAAD5Ip3+AAAADUlEQVQIHWPYsWPHfwAHfAMoI6JxIgAAAABJRU5ErkJggg==",
  animateDuration,
}) {
  const { loading, imgSrc, onError } = useLazyImg(
    imgUrl,
    placeholderUrl,
    null,
    null,
    fallbackUrl
  );
  return (
    <>
      <img
        src={imgSrc}
        onError={onError}
        style={{
          width: "100%",
          height: "100%",
          display: "block",
          objectFit: 'cover'
        }}
      />
      <BluredImage
        src={placeholderUrl}
        onError={onError}
        loading={loading}
        animateDuration={animateDuration}
        style={{
          width: "100%",
          height: "100%",
          display: "block",
        }}
      />
    </>
  );
}

LazyLoadImage.defaultProps = {};

addPropertyControls(LazyLoadImage, {
  imgUrl: {
    type: ControlType.Image,
  },
  placeholderUrl: {
    type: ControlType.Image,
    defaultValue: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAAaADAAQAAAABAAAAAQAAAAD5Ip3+AAAADUlEQVQIHWPYsWPHfwAHfAMoI6JxIgAAAABJRU5ErkJggg==",
  },
  fallbackUrl: {
    type: ControlType.Image,
    defaultValue: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAAaADAAQAAAABAAAAAQAAAAD5Ip3+AAAADUlEQVQIHWPYsWPHfwAHfAMoI6JxIgAAAABJRU5ErkJggg==",
  },
  animateDuration: {
    type: ControlType.Number,
    defaultValue: 0.28,
  },
});
